### DISCUSSION TITLE
Can I have some help understanding Uni transforms?

### DISCUSSION THREAD

**Discussion by flowt-au** [quarkusio/quarkus#28368] [Q&A] [answered]
I have a consumer like this which awaits a Vertx Future then returns the result as a `Uni<MessageOutBean>`.
```java
@ConsumeEvent("bus01F")
  public Uni<MessageOutBean> bus01F(MessagePacket messagePacket) {
    // Relay the request as an incoming message and await the future response
    Future<MessageOutBean> futureResponse = this.processIncomingMessagePacket(messagePacket, redisJsonDatasource);

    // Turn the completed future (when it IS completed) into the Uni
    return Uni.createFrom().completionStage(futureResponse.toCompletionStage());
  }
```
I understand I might be able to refactor my code to use a Uni instead of a Future, but that is for a future exploration (no pun intended).

Here is an abbreviated version of the request code method, which works.
```java
protected Uni<DataGraph> createAnEdge(String c_mkey, String p_mkey) {
  Promise<DataGraph> promise = Promise.promise();
  final Future<DataGraph> future = promise.future();

  // Build the request message packet bean
  MessagePacket messagePacket = new MessagePacket(c_mkey, p_mkey);

  // Make the request
  Uni<Message<MessageOutBean>> response = eventBus.request("bus01F", messagePacket);

  // Extract the DataGraph
  response.subscribe().with(
      messageOutBean -> {
        promise.complete(this.getDataGraphFromMessage(messageOutBean));
      },
      failure -> {
        promise.fail("create edge failure=" + failure);
      });

  // Return the Uni<DataGraph> from the future
  return Uni.createFrom().completionStage(future.toCompletionStage());
}
```
The request is being handled correctly and is returning the correct `MessageOutBean`.

The `MessageOutBean` is a generalised "packet" which contains a child bean property I wish to extract. From that child bean I want to construct another Uni, in this case a `Uni<DataGraph>`. (Different MessageOutBeans can contain different child beans but I know what the type is when I get the MessageOutBean EventBus reply because I know what type I sent in the first place).
 
So, I am receiving back a `Uni<MessageOutBean>` and I can extract the `DataGraph` as above by using another Future / promise. 

**My question is:** I think there might be a simpler way to return a `Uni<DataGraph>` using a Uni transform of `Uni<Message<MessageOutBean>>` instead of a Future. I tried a number of approaches and couldn't come up with a way that avoids using the Future. 

This is not critical since I have a working solution but I am interested in better understanding Mutiny.

Thanks,
Murray

**Comment by alex-martel** [comment, marked as answer]
I am not a mutiny expert, but hopefuly I can help a bit.

I think you can try do something like the code below. Obviously I have not tested since I do not have your code. I also know nothing about eventBus and such.

```
    protected Uni<Datagraph> createAnEdge(String c_mkey, String p_mkey) {

        // Build the request message packet bean
        MessagePacket messagePacket = new MessagePacket(c_mkey, p_mkey);

        // Make the request
        return  eventBus.request("bus01F", messagePacket).onItem().transformToUni(message -> Uni.createFrom().item(this.getDataGraphFromMessage(message)));
    }

```

You can add a `onFailure()` then do whatever you need in the failure case.

**Reply by flowt-au** [reply]
Thank you @manofthepeace ! Yes, that works. The only thing I needed to do was cast the response so I had access to the methods of the MessageOutBean in order to get the child bean inside that message:
```java
return eventBus.<MessageOutBean>request("bus01F", messagePacket).onItem().transformToUni(message -> Uni.createFrom().item(this.getDataGraphFromMessage(message)));
```
Without that typing, the `message` item from `.onItem()` is type `<Message<Object>>` even though the consumer was replying with type `Uni<MessageOutBean>`. 

Much appreciated,
Murray
